File: sdk/flutter/praxy_flutter/lib/src/realtime/realtime.dart
import 'package:praxy_core/praxy_core.dart';

import '../account_event.dart';
import '../connection_state.dart';
import 'realtime_socket.dart';

/// Query methods that filter or shape a single row's visibility — the subset worth
/// re-applying to a realtime event. Pagination/ordering methods (`limit`, `offset`,
/// `cursorAfter`, `cursorBefore`, `orderAsc`, `orderDesc`) describe how a *list* is
/// shaped, not whether one row belongs in it, so they're dropped rather than sent
/// to the single-row existence check in [PraxyRealtime.rows].
const _filterMethods = {
  'equal', 'notEqual', 'lessThan', 'lessThanEqual', 'greaterThan', 'greaterThanEqual',
  'between', 'isNull', 'isNotNull', 'startsWith', 'endsWith', 'contains', 'search',
  'and', 'or', 'select',
};

const Col<String> _idCol = Col<String>(r'$id');

/// Realtime access: `rows<T>`, `account`, `connection`, `close` — the 4-method
/// surface research/flutter-sdk.md specifies. One [RealtimeSocket] backs every
/// stream this returns, lazily created on first use.
///
/// The server's row-change event carries only `{databaseId, tableId, rowId, roles}`
/// — never the row's actual column data (`RowsService.BuildEvent`, Phase 4) — so
/// `create`/`update` events here re-fetch the row over REST before emitting. That
/// fetch also re-applies [queries] server-side (`Query.equal($id, rowId)` ANDed in),
/// which is how this stays consistent with a `list()` snapshot taken with the same
/// queries: the phase-4 report is explicit that a subscribe's `queries` field is
/// accepted but never applied as a server-side realtime filter (matching Appwrite's
/// own behaviour), so doing nothing here would leak rows the original query
/// would have excluded.
final class PraxyRealtime {
  PraxyRealtime(this._client);

  final Praxy _client;
  RealtimeSocket? _socket;

  RealtimeSocket _ensureSocket() => _socket ??= RealtimeSocket(
    endpoint: _client.endpoint,
    projectId: _client.projectId,
    ticketProvider: () async {
      final session = await _client.sessionStore.read();
      if (session == null) return null;
      try {
        return (await _client.mintRealtimeTicket()).value;
      } on PraxyException {
        return null; // fall back to an unauthenticated (guest-role) connection
      }
    },
  );

  /// A stream of [RowEvent]s for [table] — the whole table, or just [rowId] if
  /// given. [queries] narrows which creates/updates are emitted, matching what a
  /// `tables.list(table, queries: queries)` snapshot would return.
  Stream<RowEvent<T>> rows<T>(TableRef<T> table, {String? rowId, List<Query> queries = const []}) async* {
    final channel = rowId == null
        ? 'databases.${table.databaseId}.tables.${table.tableId}.rows'
        : 'databases.${table.databaseId}.tables.${table.tableId}.rows.$rowId';
    final filterQueries = [for (final q in queries) if (_filterMethods.contains(q.method)) q];

    await for (final data in _ensureSocket().subscribe([channel])) {
      final events = ((data['events'] as List?) ?? const []).cast<String>();
      final payload = (data['payload'] as Map?)?.cast<String, dynamic>() ?? const {};
      final eventRowId = payload['rowId'] as String?;
      if (eventRowId == null) continue;
      final timestamp = DateTime.tryParse(data['timestamp'] as String? ?? '') ?? DateTime.now().toUtc();
      final action = _actionOf(events);

      if (action == 'delete') {
        yield RowDeleted<T>(tableId: table.tableId, rowId: eventRowId, events: events, timestamp: timestamp);
        continue;
      }

      final row = await _fetchIfMatching(table, eventRowId, filterQueries);
      if (row == null) continue; // filtered out, no longer readable, or already gone

      yield action == 'create'
          ? RowCreated<T>(row: row, tableId: table.tableId, rowId: eventRowId, events: events, timestamp: timestamp)
          : RowUpdated<T>(row: row, tableId: table.tableId, rowId: eventRowId, events: events, timestamp: timestamp);
    }
  }

  /// The caller's own `account` channel — session/profile lifecycle events.
  Stream<AccountEvent> get account => _ensureSocket().subscribe(['account']).map(
    (data) => AccountEvent(
      events: ((data['events'] as List?) ?? const []).cast<String>(),
      channels: ((data['channels'] as List?) ?? const []).cast<String>(),
      payload: (data['payload'] as Map?)?.cast<String, dynamic>() ?? const {},
      timestamp: DateTime.tryParse(data['timestamp'] as String? ?? '') ?? DateTime.now().toUtc(),
    ),
  );

  /// Transport health — reconnects and their backoff never surface as row events.
  Stream<PraxyConnectionState> get connection => _ensureSocket().connectionStates;

  /// Tears down the shared socket and every active subscription. A later `rows`/
  /// `account` call lazily opens a fresh one.
  void close() {
    _socket?.dispose();
    _socket = null;
  }

  Future<T?> _fetchIfMatching<T>(TableRef<T> table, String rowId, List<Query> filterQueries) async {
    try {
      final page = await _client.tables.list(
        table,
        queries: [...filterQueries, Query.equal(_idCol, rowId), Query.limit(1)],
        total: false,
      );
      return page.rows.isEmpty ? null : page.rows.first;
    } on PraxyApiException {
      return null; // e.g. the table or a permission check went away mid-flight
    }
  }

  static String _actionOf(List<String> events) {
    for (final e in events) {
      if (e.endsWith('.create')) return 'create';
      if (e.endsWith('.delete')) return 'delete';
      if (e.endsWith('.update')) return 'update';
    }
    return 'update';
  }
}

File: sdk/js/packages/core/src/services/realtime.ts
import type { Praxy } from "../client.js";
import type { ConnectionState, ServerFrame, Unsubscribe } from "../realtime-socket.js";
import { RealtimeSocket } from "../realtime-socket.js";
import type { TableRef } from "../table-ref.js";

export type { ConnectionState, Unsubscribe } from "../realtime-socket.js";

export interface RowChangeEvent {
  action: "create" | "update" | "delete";
  rowId: string;
}

/**
 * A generic account/session change notification. The exact per-action grammar on this channel
 * isn't pinned down the way the row-event grammar is (`databases.{db}.tables.{tbl}.rows.{id}.{action}`)
 * — `event` carries the server's raw dot-path so a caller can still switch on it, and `payload`
 * is passed through unparsed.
 */
export interface AccountChangeEvent {
  event: string;
  payload: unknown;
}

/**
 * Client-side realtime — a Server Component can't hold a WebSocket across a request/response
 * cycle, so this only ever runs in the browser (see `@praxy/react`'s `useLiveList`). Authenticates
 * with the same JWT bridge as `@praxy/react`'s REST calls: `Praxy.mintRealtimeTicket()` exchanges
 * whatever credential this client holds for a single-use WS ticket, since the browser's native
 * `WebSocket` can't set the `X-Praxy-Session`/`X-Praxy-Key` headers a plain request would.
 *
 * 4 methods, matching `praxy_flutter`'s `PraxyRealtime`
 * (`sdk/flutter/praxy_flutter/lib/src/realtime/realtime.dart`): `rows`, `account`, `connection`, `close`.
 */
export class RealtimeService {
  private socket: RealtimeSocket | null = null;

  constructor(private readonly client: Praxy) {}

  /**
   * Subscribes to row changes on a table (or one row, if `rowId` is given). Delivers only
   * `{action, rowId}` — the server's row-change event never carries column data — so a caller
   * (typically `@praxy/react`'s `useLiveList`) re-fetches via `tables.get`/`tables.list` on each event.
   */
  rows<T>(table: TableRef<T>, listener: (event: RowChangeEvent) => void, options: { rowId?: string } = {}): Unsubscribe {
    const channel = options.rowId
      ? `databases.${table.databaseId}.tables.${table.tableId}.rows.${options.rowId}`
      : `databases.${table.databaseId}.tables.${table.tableId}.rows`;

    const pattern = new RegExp(
      `^databases\\.${escapeRegExp(table.databaseId)}\\.tables\\.${escapeRegExp(table.tableId)}\\.rows\\.([^.]+)\\.(create|update|delete)$`,
    );

    return this.ensureSocket().subscribeChannel(channel, (frame) => {
      const change = parseRowChange(frame, pattern);
      if (change) listener(change);
    });
  }

  /** Subscribes to the caller's own account/session events (server-rewritten to `account.<userId>`). */
  account(listener: (event: AccountChangeEvent) => void): Unsubscribe {
    return this.ensureSocket().subscribeChannel("account", (frame) => {
      const data = frame.data as { events?: string[]; payload?: unknown } | undefined;
      const event = data?.events?.[0];
      if (event) listener({ event, payload: data?.payload });
    });
  }

  /** The shared socket's own connection lifecycle — replays the current state to a new listener immediately. */
  connection(listener: (state: ConnectionState) => void): Unsubscribe {
    return this.ensureSocket().subscribeConnectionState(listener);
  }

  /** Disposes the shared socket. A later `rows`/`account`/`connection` call lazily reopens one. */
  close(): void {
    this.socket?.close();
  }

  private ensureSocket(): RealtimeSocket {
    if (!this.socket) {
      this.socket = new RealtimeSocket({
        wsUrl: toWebSocketUrl(this.client.endpoint, this.client.projectId),
        mintTicket: async () => {
          try {
            return (await this.client.mintRealtimeTicket()).ticket;
          } catch {
            return null; // fall back to an unauthenticated (guest-role) connection
          }
        },
      });
    }
    return this.socket;
  }
}

function parseRowChange(frame: ServerFrame, pattern: RegExp): RowChangeEvent | null {
  const events = (frame.data as { events?: string[] } | undefined)?.events ?? [];
  for (const event of events) {
    const match = pattern.exec(event);
    if (match) return { rowId: match[1]!, action: match[2] as RowChangeEvent["action"] };
  }
  return null;
}

function toWebSocketUrl(endpoint: string, projectId: string): string {
  const url = new URL(endpoint);
  url.protocol = url.protocol === "https:" ? "wss:" : "ws:";
  url.pathname = url.pathname.replace(/\/+$/, "") + "/v1/realtime";
  url.search = `?project=${encodeURIComponent(projectId)}`;
  return url.toString();
}

function escapeRegExp(value: string): string {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
